src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -321,107 +300,107 @@ private bool TryLookupLengthOrCount(SyntaxNode node, TypeSymbol inputType, out P
+                elementType = CreateErrorType();
             }
 
-            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node.Subpatterns, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
-            return new BoundListPatternClause(node, subpatterns, sawSlice, indexerAccess, indexerSymbol, hasErrors);
+            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node.Patterns, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
+            TypeSymbol narrowedType = inputType.StrippedType();
+            BindPatternDesignation(
+                node.Designation, TypeWithAnnotations.Create(narrowedType, NullableAnnotation.NotAnnotated), inputValEscape, permitDesignations, node, diagnostics,
+                ref hasErrors, out Symbol? variableSymbol, out BoundExpression? variableAccess);
+            return new BoundListPattern(node, subpatterns, sawSlice, lengthProperty, indexerAccess, indexerSymbol, variableSymbol, variableAccess, inputType, narrowedType: narrowedType, hasErrors);
         }
 
-        private bool TryFindIndexerOrIndexerPattern(
+        private bool TryPerformPatternIndexerLookup(
             SyntaxNode syntax, TypeSymbol receiverType, bool argIsIndex,
             out BoundIndexerAccess? indexerAccess,
             out Symbol? patternSymbol,
+            [NotNullWhen(true)] out PropertySymbol? lengthProperty,
             BindingDiagnosticBag diagnostics)
         {
+            indexerAccess = null;
             patternSymbol = null;
-            var lookupResult = LookupResult.GetInstance();
-            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
-            if (TryPerformPatternIndexerLookup(syntax, receiverType, lookupResult, argIsIndex, out indexerAccess, diagnostics, ref useSiteInfo) ||
-                TryFindIndexOrRangeIndexerPattern(syntax, lookupResult, receiverOpt: null, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo, forPattern: true))
-            {
-                diagnostics.Add(syntax, useSiteInfo);
-                lookupResult.Free();
-                return true;
-            }
-
-            lookupResult.Free();
-            return false;
-        }
-
-        private bool TryPerformPatternIndexerLookup(
-            SyntaxNode syntax,
-            TypeSymbol receiverType,
-            LookupResult lookupResult,
-            bool argIsIndex,
-            [NotNullWhen(true)] out BoundIndexerAccess? result,
-            BindingDiagnosticBag diagnostics,
-            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
-        {
-            result = null;
+            bool found;
 
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            var bindingDiagnostics = BindingDiagnosticBag.Create(diagnostics);
```

Comment:
`GetInstance`?

Reply:
The argument is a "template", it doesn't actually pull diagnostics.